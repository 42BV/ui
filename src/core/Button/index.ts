export { Button, Props, ButtonSize, ButtonIconPosition } from './Button';
