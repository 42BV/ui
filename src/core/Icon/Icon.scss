/**
* Make material-icons within Bootstrap great again!
*/
.input-group {
  &-append,
  &-prepend {
    i {
      @extend .input-group-text;
    }

    .btn > i.material-icons {
      border: 0;
      background: none;
      color: #fff;
      padding: 0;
    }
  }
}

i.material-icons {
  user-select: none;
}

.nav-link {
  i.material-icons {
    vertical-align: middle;
  }
}
