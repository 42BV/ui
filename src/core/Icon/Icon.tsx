import React from 'react';
import classNames from 'classnames';

import { Color, useHover } from '../..';
import IconType from './types';

export type Props = {
  /**
   * The material icon you want to render.
   */
  icon: IconType;

  /**
   * Optional color you want the Icon to have.
   */
  color?: Color;

  /**
   * Optional color you want the Icon to have when it is hovered.
   * When `disabled` is true, hoverColor does not work.
   * This will only be used when onClick is defined.
   */
  hoverColor?: Color;

  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   */
  className?: string;

  /**
   * Optional id when using controlled tooltips.
   */
  id?: string;

  /**
   * Optional onClick event for when the Icon is clicked.
   */
  onClick?: (event: React.MouseEvent<HTMLElement>) => void;

  /**
   * Optionally whether the button is disabled
   *
   * Defaults to `false`
   */
  disabled?: boolean;

  /**
   * Optionally the size of the icon in pixels.
   *
   * Defaults to `24px`
   */
  size?: number;
};

/**
 * The Icon is a small wrapper around a material design icon.
 *
 * Useful for making sure your icon is typesafe via TypeScript.
 */
export default function Icon({
  className,
  onClick,
  icon,
  id,
  color,
  hoverColor,
  disabled,
  size
}: Props) {
  const [hover, hoverEvents] = useHover();

  const colorCssClasses = {};
  if (color) {
    colorCssClasses[`text-${color}`] = (hoverColor && !hover) || disabled || onClick === undefined;
  }
  if (hoverColor) {
    colorCssClasses[`text-${hoverColor}`] = hover && !disabled && onClick !== undefined;
  }

  const classes = classNames(
    'icon',
    className,
    'material-icons',
    colorCssClasses,
    {
      clickable: !disabled && onClick,
      'icon--disabled': disabled
    }
  );

  const style = size ? { fontSize: size } : undefined;

  return (
    <i
      id={id}
      style={style}
      onClick={(event) => {
        if (!disabled && onClick) {
          onClick(event);
        }
      }}
      className={classes}
      {...hoverEvents}
    >
      {icon}
    </i>
  );
}
