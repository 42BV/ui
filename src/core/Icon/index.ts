export { Icon } from './Icon';
export { IconType, icons } from './types';
