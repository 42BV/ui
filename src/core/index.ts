export { default as Avatar } from './Avatar/Avatar';
export { default as AvatarStack } from './Avatar/AvatarStack';
export { default as Breadcrumbs } from './Breadcrumbs/Breadcrumbs';
export { default as Button } from './Button/Button';
export { default as useShowSpinner } from './Button/useShowSpinner';
export { default as ConfirmButton } from './ConfirmButton/ConfirmButton';
export { default as ContentState } from './ContentState/ContentState';
export { default as FlashMessage } from './FlashMessage/FlashMessage';
export { Icon, IconType } from './Icon/index';
export { default as InfoBadge } from './InfoBadge/InfoBadge';
export { default as LoadingPage } from './LoadingPage/LoadingPage';
export { default as MoreOrLess } from './MoreOrLess/MoreOrLess';
export { default as NavigationItem } from './NavigationItem/NavigationItem';
export { default as Pagination } from './Pagination/Pagination';
export { default as ProgressStepper } from './ProgressStepper/ProgressStepper';
export { default as Spinner } from './Spinner/Spinner';
export { default as SubmitButton } from './SubmitButton/SubmitButton';
export { default as Tag } from './Tag/Tag';
