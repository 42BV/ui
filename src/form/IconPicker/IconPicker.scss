.icon-picker {
  &-clear {
    font-size: 12px;
    display: inline;
    cursor: pointer;
    text-decoration: underline !important;
    position: relative;
    top: -8px;
    margin: 0 8px 0 10px;
  }
}
