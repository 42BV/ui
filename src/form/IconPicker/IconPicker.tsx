import React, { useState } from 'react';
import {
  Button,
  FormGroup,
  Label,
  Card,
  CardHeader,
  CardBody
} from 'reactstrap';
import classNames from 'classnames';

import withJarb from '../withJarb/withJarb';
import { Icon, icons, IconType } from '../../core/Icon';
import Pager from '../../core/Pager/Pager';
import { doBlur } from '../utils';
import SearchInput from '../../core/SearchInput/SearchInput';
import { pageOf } from '../../utilities/page/page';
import ContentState from '../../core/ContentState/ContentState';
import { t } from '../../utilities/translation/translation';
import Tooltip from '../../core/Tooltip/Tooltip';
import Popover from '../../core/Popover/Popover';
import TextButton from '../../core/TextButton/TextButton';
import { FieldCompatible } from '../types';

type Text = {
  /**
   * No results title to show when no icon could be found for that
   * query.
   */
  noResultsTitle?: string;

  /**
   * No results subtitle to show when no icon could be found for that
   * query.
   */
  noResultsSubtitle?: string;

  /**
   * The text of the clear button
   */
  clear?: string;
};

type Props = FieldCompatible<IconType, IconType | undefined> & {
  /**
   * Optionally customized text within the component.
   * This text should already be translated.
   */
  text?: Text;

  /**
   * Optionally the icon to display on the button to open the icon picker.
   */
  icon?: IconType;

  /**
   * Whether or not to show a "clear" button.
   *
   * Defaults to `true`
   */
  canClear?: boolean;
};

/**
 * IconPicker is a form element which allows the user to select one
 * of the material design icons. It is a popover which shows all
 * the icons and lets the user select one.
 */
export default function IconPicker(props: Props) {
  const {
    label,
    icon,
    value,
    color,
    placeholder,
    className,
    error,
    onChange,
    onBlur,
    text = {},
    canClear = true
  } = props;

  const [isOpen, setIsOpen] = useState(false);
  const [pageNumber, setPageNumber] = useState(1);
  const [query, setQuery] = useState('');

  const filteredIcons = icons.filter((icon) =>
    icon.toLowerCase().startsWith(query.toLowerCase())
  );

  const iconsPage = pageOf(filteredIcons, pageNumber, 36);

  function onIconSelected(icon?: IconType) {
    onChange(icon);
    doBlur(onBlur);
    setIsOpen(false);
  }

  function onSearch(query: string) {
    setQuery(query);
    setPageNumber(1);
  }

  const isEmpty = filteredIcons.length === 0;

  const classes = classNames('icon-picker', className);

  return (
    <SearchInput defaultValue={query} onChange={onSearch} debounce={0}>
      {(searchInput, searchInputApi) => (
        <FormGroup className={classes} color={color}>
          {label ? <Label>{label}</Label> : null}

          <div className="d-flex">
            {value ? (
              <div className="d-flex justify-content-between">
                <Icon
                  id="icon-picker-value"
                  className="pt-2 me-3"
                  icon={value}
                />
                {canClear ? (
                  <TextButton
                    onClick={() => onIconSelected(undefined)}
                    className="me-3"
                  >
                    {t({
                      key: 'IconPicker.CLEAR',
                      fallback: 'Clear',
                      overrideText: text.clear
                    })}
                  </TextButton>
                ) : null}
              </div>
            ) : null}
            <Popover
              placement="bottom"
              isOpen={isOpen}
              target={
                <Button
                  type="button"
                  color="primary"
                  onClick={() => {
                    if (isOpen) {
                      searchInputApi.setValue('');
                    }

                    setIsOpen(!isOpen);
                  }}
                >
                  {icon ? (
                    <Icon icon={icon} className="me-2 align-bottom" />
                  ) : null}
                  {placeholder}
                </Button>
              }
            >
              <Card>
                <CardHeader>{searchInput}</CardHeader>
                <CardBody>
                  <div style={{ width: 250, height: 250 }}>
                    {!isEmpty ? (
                      iconsPage.content.map((icon) => (
                        <span key={icon} className="d-inline-block">
                          <Tooltip placement="top" distance={18} content={icon}>
                            <Icon
                              className="m-2"
                              icon={icon}
                              onClick={() => onIconSelected(icon)}
                            />
                          </Tooltip>
                        </span>
                      ))
                    ) : (
                      <ContentState
                        mode="no-results"
                        title={t({
                          key: 'IconPicker.NO_RESULTS.TITLE',
                          fallback: 'No icons found',
                          overrideText: text.noResultsTitle
                        })}
                        subTitle={t({
                          key: 'IconPicker.NO_RESULTS.SUBTITLE',
                          fallback:
                            'No icons were found, please try again with a different query.',
                          overrideText: text.noResultsSubtitle
                        })}
                      />
                    )}
                  </div>

                  <div className="my-2 text-center">
                    <Pager page={iconsPage} onChange={setPageNumber} />
                  </div>
                </CardBody>
              </Card>
            </Popover>
          </div>

          {error}
        </FormGroup>
      )}
    </SearchInput>
  );
}

/**
 * Variant of the IconPicker which can be used in a Jarb context.
 */
export const JarbIconPicker = withJarb<IconType, IconType, Props>(IconPicker);
