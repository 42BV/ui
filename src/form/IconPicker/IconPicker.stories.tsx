import React, { useState } from 'react';
import { storiesOf } from '@storybook/react';

import { IconPicker, JarbIconPicker } from './IconPicker';
import { FinalForm, JarbFormElementDependencies } from '../story-utils';
import { Card, Icon, IconType, Tooltip } from '../../';

function is3DRotation(value: IconType) {
  return value === '3d_rotation' ? undefined : 'Not "3d_rotation"';
}

storiesOf('Form/IconPicker', module)
  .addParameters({ component: IconPicker })
  .add('basic', () => {
    const [ value, setValue ] = useState<IconType | undefined>(undefined);

    return (
      <div>
        <Card className="m-2">
          <IconPicker
            id="icon"
            label="Icon"
            placeholder="Please select your icon"
            value={value}
            onChange={setValue}
          />
        </Card>
      </div>
    );
  })
  .add('invisible label', () => {
    const [ value, setValue ] = useState<IconType | undefined>(undefined);

    return (
      <Card className="m-2">
        <IconPicker
          id="icon"
          label="Icon"
          hiddenLabel={true}
          placeholder="Please select your icon"
          value={value}
          onChange={setValue}
        />
      </Card>
    );
  })
  .add('with custom label', () => {
    const [ value, setValue ] = useState<IconType | undefined>(undefined);

    return (
      <div>
        <Card className="m-2">
          <IconPicker
            id="icon"
            label={
              <div className="d-flex justify-content-between">
                <span>Icon</span>
                <Tooltip
                  className="ms-1"
                  content="The icon will be used as an avatar"
                >
                  <Icon icon="info" />
                </Tooltip>
              </div>
            }
            placeholder="Please select your icon"
            value={value}
            onChange={setValue}
          />
        </Card>
      </div>
    );
  })
  .add('with icon', () => {
    const [ value, setValue ] = useState<IconType | undefined>(undefined);

    return (
      <div>
        <Card className="m-2">
          <IconPicker
            id="icon"
            label="Icon"
            placeholder="Please select your icon"
            icon="colorize"
            value={value}
            onChange={setValue}
          />
        </Card>
      </div>
    );
  })
  .add('without clear button', () => {
    const [ value, setValue ] = useState<IconType | undefined>(undefined);

    return (
      <div>
        <Card className="m-2">
          <IconPicker
            id="icon"
            label="Icon"
            placeholder="Please select your icon"
            value={value}
            onChange={setValue}
            canClear={false}
          />
        </Card>
      </div>
    );
  })
  .add('jarb', () => {
    return (
      <>
        <JarbFormElementDependencies />
        <FinalForm>
          <JarbIconPicker
            id="icon"
            name="icon"
            label="Icon"
            placeholder="Please select your icon"
            validators={[ is3DRotation ]}
            jarb={{
              validator: 'Hero.icon',
              label: 'Icon'
            }}
          />
        </FinalForm>
      </>
    );
  });
