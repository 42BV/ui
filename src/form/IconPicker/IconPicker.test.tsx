import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';

import { IconPicker } from './IconPicker';
import { IconType } from '../../core/Icon';

describe('Component: IconPicker', () => {
  function setup({
    value,
    hasLabel,
    hasIcon,
    canClear
  }: {
    value?: IconType;
    hasLabel?: boolean;
    hasIcon?: boolean;
    canClear?: boolean;
  }) {
    const onChangeSpy = jest.fn();
    const onBlurSpy = jest.fn();

    const props = {
      id: 'bestFriend',
      name: 'bestFriend',
      label: hasLabel ? 'Best Friend' : undefined,
      placeholder: 'Select your best friend',
      icon: hasIcon ? 'face' as IconType : undefined,
      value,
      onChange: onChangeSpy,
      onBlur: onBlurSpy,
      error: 'Some error',
      canClear
    };

    const { container, asFragment, rerender } = render(
      <IconPicker color="success" {...props} />
    );

    return {
      container,
      props,
      rerender,
      asFragment,
      onBlurSpy,
      onChangeSpy
    };
  }

  describe('ui', () => {
    test('with selected value', () => {
      const { container } = setup({ value: '3d_rotation' });
      expect(container).toMatchSnapshot();
    });

    test('without selected value', () => {
      const { container } = setup({});
      expect(container).toMatchSnapshot();
    });

    test('no results', async () => {
      expect.assertions(1);

      setup({
        value: undefined
      });

      fireEvent.click(screen.getByText('Select your best friend'));
      await userEvent.type(screen.getByRole('searchbox'), 'asdlfjalsdjflajsdlf');

      expect(screen.queryByText('No icons found')).toBeInTheDocument();
    });

    test('with label', () => {
      const { container } = setup({ hasLabel: true });
      expect(screen.queryByText('Best Friend')).toBeInTheDocument();
      expect(container).toMatchSnapshot();
    });

    test('without label', () => {
      setup({ hasLabel: false });
      expect(screen.queryByText('Best Friend')).not.toBeInTheDocument();
    });

    test('with icon', () => {
      setup({ hasIcon: true });
      expect(screen.getByText('face')).toBeInTheDocument();
    });

    test('without clear button', () => {
      setup({ value: '3d_rotation', canClear: false });
      expect(screen.queryByText('clear')).not.toBeInTheDocument();
    });
  });

  describe('events', () => {
    it('should open the popover when the select button is clicked', () => {
      const setIsOpenSpy = jest.fn();
      jest.spyOn(React, 'useState')
        .mockReturnValueOnce([false, setIsOpenSpy]);

      setup({
        value: undefined,
        hasIcon: false
      });

      expect(setIsOpenSpy).toHaveBeenCalledTimes(0);

      fireEvent.click(screen.getByText('Select your best friend'));

      expect(setIsOpenSpy).toHaveBeenCalledTimes(1);
    });

    it('should filter the icons when the user searches and go back to the first page', () => {
      setup({
        value: undefined,
        hasIcon: true
      });

      fireEvent.click(screen.getByText('face'));
      fireEvent.change(screen.getByRole('searchbox'), { target: { value: 'home' } });

      expect(screen.queryAllByText('home').length).toBe(0);
    });

    it('should when the user moves to another page load the new page', () => {
      setup({
        value: undefined
      });

      fireEvent.click(screen.getByText('Select your best friend'));

      // The first icon on the first page is the 3d_rotation
      expect(screen.getByText('10k')).toBeInTheDocument();

      fireEvent.click(screen.getByText('arrow_forward'));

      expect(screen.queryByText('10k')).not.toBeInTheDocument();
      expect(screen.getByText('4mp')).toBeInTheDocument();
    });

    it('should close the popover and call onChange when the user selects an icon', () => {
      const setIsOpenSpy = jest.fn();
      jest.spyOn(React, 'useState')
        .mockReturnValueOnce([true, setIsOpenSpy]);

      const { onBlurSpy, onChangeSpy } = setup({
        value: undefined
      });

      fireEvent.click(screen.getByText('10k'));

      expect(onChangeSpy).toHaveBeenCalledTimes(1);
      expect(onChangeSpy).toHaveBeenCalledWith('10k');

      expect(onBlurSpy).toHaveBeenCalledTimes(1);

      expect(setIsOpenSpy).toHaveBeenCalledTimes(1);
      expect(setIsOpenSpy).toHaveBeenCalledWith(false);
    });

    it('should when the user clicks "clear" reset the icon', () => {
      const { onBlurSpy, onChangeSpy } = setup({ value: 'home' });

      fireEvent.click(screen.getByText('Clear'));

      expect(onChangeSpy).toHaveBeenCalledTimes(1);
      expect(onChangeSpy).toHaveBeenCalledWith(undefined);

      expect(onBlurSpy).toHaveBeenCalledTimes(1);
    });

    it('should when the popover closes clear the query', () => {
      const setQuery = jest.fn();
      jest.spyOn(React, 'useState')
        .mockReturnValueOnce([true, jest.fn()])
        .mockReturnValueOnce([1, jest.fn()])
        .mockReturnValueOnce(['home', setQuery]);

      setup({
        value: undefined
      });

      fireEvent.click(screen.getByText('Select your best friend'));

      expect(setQuery).toHaveBeenCalledTimes(1);
      expect(setQuery).toHaveBeenCalledWith('');
    });
  });

  describe('value changes', () => {
    test('becomes empty', () => {
      const { props, rerender } = setup({
        value: 'home'
      });

      expect(screen.getByText('home')).toBeInTheDocument();

      const newProps = {
        ...props,
        value: undefined
      };

      rerender(
        <IconPicker color="success" {...newProps} />
      );

      expect(screen.queryByText('home')).not.toBeInTheDocument();
    });

    test('becomes filled', () => {
      const { props, rerender } = setup({
        value: undefined
      });

      expect(screen.queryByText('home')).not.toBeInTheDocument();

      const newProps = {
        ...props,
        value: 'home' as IconType
      };

      rerender(
        <IconPicker color="success" {...newProps} />
      );

      expect(screen.getByText('home')).toBeInTheDocument();
    });
  });
});
