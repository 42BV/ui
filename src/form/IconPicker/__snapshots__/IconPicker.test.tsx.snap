// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: IconPicker ui with label 1`] = `
<div>
  <div
    class="search-input"
  >
    <div
      class="icon-picker mb-3"
      color="success"
    >
      <label
        class="form-label"
      >
        Best Friend
      </label>
      <div
        class="d-flex"
      >
        <button
          aria-describedby="bestFriend"
          class="btn btn-primary"
          type="button"
        >
          Select your best friend
        </button>
      </div>
      Some error
    </div>
  </div>
</div>
`;

exports[`Component: IconPicker ui with selected value 1`] = `
<div>
  <div
    class="search-input"
  >
    <div
      class="icon-picker mb-3"
      color="success"
    >
      <div
        class="d-flex"
      >
        <div
          class="d-flex justify-content-between"
        >
          <i
            class="icon pt-2 me-3 material-icons"
            id="icon-picker-value"
          >
            3d_rotation
          </i>
          <u
            class="align-self-center clickable fw-lighter me-3"
            role="button"
          >
            Clear
          </u>
        </div>
        <button
          aria-describedby="1"
          class="btn btn-primary"
          type="button"
        >
          Select your best friend
        </button>
      </div>
      Some error
    </div>
  </div>
</div>
`;

exports[`Component: IconPicker ui without selected value 1`] = `
<div>
  <div
    class="search-input"
  >
    <div
      class="icon-picker mb-3"
      color="success"
    >
      <div
        class="d-flex"
      >
        <button
          aria-describedby="3"
          class="btn btn-primary"
          type="button"
        >
          Select your best friend
        </button>
      </div>
      Some error
    </div>
  </div>
</div>
`;
